docs/source/nlp/nemo_megatron/retro/retro_model.rst

```diff
@@ -1,469 +1,170 @@
-        --learned_index=/result/pubmed_faiss_learn.index \
-        --shard_index_input=/result/pubmed_faiss_shard \
-        --output_file=/result/pubmed_faiss_final.index
+The main output of this stage are:
 
-Step 4: Build KNN index
-^^^^^^^^^^^^^^^^^^^^^^^
+* ``/db/merged/train.hdf5``: the database containing all processed and chunked text.
+* ``/db/merged/sampled.hdf5``: the database containing a small portion of all chunks, only used for training the index in the next stage.
 
-During training, it is inefficient to run a query to find the K-nearest neighbor chunk IDs for each training data point. 
-This can be pre-calculated by building a KNN index before training. The KNN index maps the training data chunk IDs to the K-nearest neighbor chunk IDs 
-in the retrieval data. As with building the Faiss index, this process is divided into two steps.
+Build Index for Similarity Search
+#################################
 
-Following is the KNN index data format:
+The second stage is to build a search index using Faiss, a library for efficient similarity search. The index is trained on a subset of the chunks ``sampled.hdf5`` from the database. After training, all chunks are added to the index to enable querying. The index accepts 1-D floating point vectors, so chunks must be embedded using Bert embeddings before they can be added to the index. Particularly, the stage is comprised of two sub-stages:
 
-.. list-table::
-   :widths: 25 25 25 25 45
+    \- Extract BERT embeddings from the sampled chunk database (``sampled.hdf5``) and use them to train a Faiss index.
 
-   * - 'KNNRETM\x00\x00' (header 9 bytes)
-     - 1 (version 8 byte)
-     - K number of neighbors (8 byte)
-     - Number chunks (8 byte)
-     - Map to K retrieval data chunk IDs, shape (number_chunks, K) ( int64 array)
+    \- Extract BERT embeddings for each chunk in the all chunks database (``train.hdf5``) and add them to the trained Faiss index.
 
-Step 4.1: Build KNN sharding index
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The main output of this stage are:
 
-The KNN index is built using the memory-mapped training data created by the ``preprocess_data_for_megatron.py`` script and the Faiss index 
-file for the retrieval data built by the ``build_retrieval_index.py`` script.
+* ``/index/<RETRO_INDEX_TYPE>/<RETRO_INDEX_STR>/added.faissindex``: the trained index, with all chunks in the database added to it
 
-An example script is:
+Query Pretraining Neighbors
+###########################
 
-.. code-block:: bash
+For training RETRO, to speed up the Retro pretraining process, we will pre-retrieve neighbors for all training samples instead of retrieving them on-the-fly. In this stage, the pretraining datasets are processed to find and save k-nearest neighbors for each chunk in each sample. The neighbors are saved to disk and labeled with unique properties to ensure they match the pretraining configuration. Query-time hyperparameters can be tuned to improve the quality of the neighbors.
```

Comment:
To speed up the RETRO pretraining process, you retrieve neighbors for all training samples ahead of time instead of retrieving them on-the-fly. In this stage, the pretraining datasets are processed to find and save k-nearest neighbors for each chunk in each sample. The neighbors are saved to disk and labeled with unique properties to ensure they match the pretraining configuration. Query-time hyperparameters can be tuned to improve the quality of the neighbors.